File: ChessEngine/bishop.py
bishop_offsets = [(1, 1), (1, -1), (-1, 1), (-1, -1)]

def check_bishop(current_board: dict,moves_list: list, value="bishop"):
    """
    The function `check_bishop` iterates through a chess board to find valid moves for bishops and adds
    them to a list.
    
    :param current_board: The function `check_bishop` seems to be checking the possible moves for
    bishops on a chess board based on the current board state. It iterates over the board positions and
    checks for valid diagonal moves for both white ('w') and black ('b') bishops
    :type current_board: dict
    :param moves_list: The `moves_list` parameter is a dictionary that contains lists of moves for each
    player. The keys in the dictionary represent the player color ('w' for white and 'b' for black), and
    the values are lists of tuples representing the moves
    :type moves_list: list
    :param value: The `value` parameter in the `check_bishop` function is used to specify the type of
    piece being checked. In this case, the function is checking for bishops on the chessboard. The
    default value for `value` is set to "bishop", indicating that the function will check for bishops,
    defaults to bishop (optional)
    """
    chain = 1 
    check = True
    
    for k, v in current_board.items():
        check = True
        if v == "B":
            for offset in bishop_offsets:
                check = True
                chain = 1
                while check:
                    if (k[0] + chain * offset[0], k[1] + chain * offset[1]) in current_board and (current_board[(k[0] + chain * offset[0], k[1] + chain * offset[1])].islower() or current_board[(k[0] + \
                            chain * offset[0], k[1] + chain * offset[1])] == ""):
                        if (k[0] + chain * offset[0], k[1] + chain * offset[1]) in current_board and current_board[(k[0] + chain * offset[0], k[1] + chain * offset[1])].isupper():
                            check = False
                            break
                            
                        chain += 1
                        
                        if check:
                            moves_list['b'].append((k, (k[0] + (chain * offset[0] - offset[0]), k[1] + (chain * offset[1] - offset[1]))))
                        if (k[0] + chain * offset[0], k[1] + chain * offset[1]) in current_board and current_board[(k[0] + chain * offset[0], k[1] + chain * offset[1])].islower():
                            check = False
                            chain += 1
                            moves_list['b'].append((k, (k[0] + (chain * offset[0] - offset[0] - offset[0]), k[1] + (chain * offset[1] - offset[1] - offset[1]))))
                    else:
                        check = False
        if v == "b":
            for offset in bishop_offsets:
                check = True
                chain = 1
                while check:
                    if (k[0] + chain * offset[0], k[1] + chain * offset[1]) in current_board and (current_board[(k[0] + chain * offset[0], k[1] + chain * offset[1])].isupper() or current_board[(k[0] + \
                            chain * offset[0], k[1] + chain * offset[1])] == ""):
                        if (k[0] + chain * offset[0], k[1] + chain * offset[1]) in current_board and current_board[(k[0] + chain * offset[0], k[1] + chain * offset[1])].islower():
                            check = False
                            break
                        chain += 1
                        
                        if check:
                            moves_list['w'].append((k, (k[0] + (chain * offset[0] - offset[0]), k[1] + (chain * offset[1] - offset[1]))))
                            
                        if (k[0] + chain * offset[0] - offset[0], k[1] + chain * offset[1] - offset[1]) in current_board and current_board[(k[0] + chain * offset[0] - offset[0], k[1] + chain * offset[1] - offset[1])].isupper():
                            
                            check = False
                            chain += 1
                            moves_list['w'].append((k, (k[0] + (chain * offset[0] - offset[0] - offset[0]), k[1] + (chain * offset[1] - offset[1] - offset[1]))))
                    else:
                        check = False

File: ChessEngine/chessEngineCPP/bishop.cpp
#include "bishop.h"
#include <map>
#include <vector>
#include <utility>
#include <ctype.h>
#include <iostream>

const std::vector<std::pair<int, int>> bishop_offsets = {
    {1, 1}, {1, -1}, {-1, 1}, {-1, -1}
};

void check_bishop(const std::map<std::pair<int, int>, char>& current_board, 
                  std::map<char, std::vector<std::pair<std::pair<int, int>, std::pair<int, int>>>>& moves_list) 
{
    for (const auto& piece : current_board) {
        std::pair<int, int> position = piece.first;
        char piece_type = piece.second;
        int chain = 1;
        
        // Check for white and black bishops
        if (piece_type == 'B' || piece_type == 'b') {
            for (const auto& offset : bishop_offsets) {
                chain = 1;
                bool check = true;
                
                while (check) {
                    std::pair<int, int> target_position = {
                        position.first + chain * offset.first,
                        position.second + chain * offset.second
                    };
                    
                    // If target position is on the board
                    if (current_board.find(target_position) != current_board.end()) {
                        char target_piece = current_board.at(target_position);
                        if ((piece_type == 'B' && isupper(target_piece)) ||
                            (piece_type == 'b' && islower(target_piece))) {
                            check = false;
                            break;
                        }
                        // Determine if it's a valid move for black or white
                        if (piece_type == 'B' && (target_piece == ' ' || islower(target_piece))) {
                            moves_list['b'].push_back({position, target_position});
                        } else if (piece_type == 'b' && (target_piece == ' ' || isupper(target_piece))) {
                            moves_list['w'].push_back({position, target_position});
                        }
                        
                        // Stop if there's an enemy piece to capture
                        if ((piece_type == 'B' && islower(target_piece)) ||
                            (piece_type == 'b' && isupper(target_piece))) {
                            check = false;
                        }
                        
                    chain++;
                    } else {
                        check = false;
                    }
                }
            }
        }
    }
}
